cni/pkg/install-cni/pkg/install/cniconfig.go

```diff
@@ -0,0 +1,325 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package install
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/containernetworking/cni/libcni"
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+
+	"istio.io/istio/cni/pkg/install-cni/pkg/config"
+	"istio.io/istio/cni/pkg/install-cni/pkg/util"
+	"istio.io/pkg/log"
+)
+
+type pluginConfig struct {
+	mountedCNINetDir string
+	cniConfName      string
+	chainedCNIPlugin bool
+}
+
+type cniConfigTemplate struct {
+	cniNetworkConfigFile string
+	cniNetworkConfig     string
+}
+
+type cniConfigVars struct {
+	cniNetDir          string
+	kubeconfigFilename string
+	logLevel           string
+	k8sServiceHost     string
+	k8sServicePort     string
+	k8sNodeName        string
+}
+
+func getPluginConfig(cfg *config.Config) pluginConfig {
+	return pluginConfig{
+		mountedCNINetDir: cfg.MountedCNINetDir,
+		cniConfName:      cfg.CNIConfName,
+		chainedCNIPlugin: cfg.ChainedCNIPlugin,
+	}
+}
+
+func getCNIConfigTemplate(cfg *config.Config) cniConfigTemplate {
+	return cniConfigTemplate{
+		cniNetworkConfigFile: cfg.CNINetworkConfigFile,
+		cniNetworkConfig:     cfg.CNINetworkConfig,
+	}
+}
+
+func getCNIConfigVars(cfg *config.Config) cniConfigVars {
+	return cniConfigVars{
+		cniNetDir:          cfg.CNINetDir,
+		kubeconfigFilename: cfg.KubeconfigFilename,
+		logLevel:           cfg.LogLevel,
+		k8sServiceHost:     cfg.K8sServiceHost,
+		k8sServicePort:     cfg.K8sServicePort,
+		k8sNodeName:        cfg.K8sNodeName,
+	}
+}
+
+func createCNIConfigFile(ctx context.Context, cfg *config.Config, saToken string) (string, error) {
+	cniConfig, err := readCNIConfigTemplate(getCNIConfigTemplate(cfg))
+	if err != nil {
+		return "", err
+	}
+
+	cniConfig = replaceCNIConfigVars(cniConfig, getCNIConfigVars(cfg), saToken)
+
+	return writeCNIConfig(ctx, cniConfig, getPluginConfig(cfg))
+}
+
+func readCNIConfigTemplate(template cniConfigTemplate) ([]byte, error) {
+	if fileutil.Exist(template.cniNetworkConfigFile) {
+		cniConfig, err := ioutil.ReadFile(template.cniNetworkConfigFile)
+		if err != nil {
+			return nil, err
+		}
+		log.Infof("Using CNI config template from %s", template.cniNetworkConfigFile)
+		return cniConfig, nil
+	}
+
+	if len(template.cniNetworkConfig) > 0 {
+		log.Infof("Using CNI config template from CNI_NETWORK_CONFIG environment variable.")
+		return []byte(template.cniNetworkConfig), nil
+	}
+
+	return nil, errors.New("need CNI_NETWORK_CONFIG or CNI_NETWORK_CONFIG_FILE to be set")
+}
+
+func replaceCNIConfigVars(cniConfig []byte, vars cniConfigVars, saToken string) []byte {
+	cniConfigStr := string(cniConfig)
+
+	cniConfigStr = strings.ReplaceAll(cniConfigStr, "__LOG_LEVEL__", vars.logLevel)
```

Comment:
Is there any reason not to use go templates here, rather than string replacement?

Reply:
I know that go templates allow you to change the delimiters, but I think it still requires each field to be preceded with a period right? 
Also the CNI config template is passed in through environment variables (`CNI_NETWORK_CONFIG` or `CNI_NETWORK_CONFIG_FILE`) which are configured in the charts, but could users define their own custom yaml files to set these environment variables? ([cni charts directory](https://github.com/istio/istio/tree/master/manifests/charts/istio-cni))
If so for both questions, then I was thinking that changing the format could break some existing configurations, so I left it as is.

Reply:
Sorry if it wasn't clear -- I was suggesting that we modify the chart yaml to escape additional go template rules that the installer can set, e.g. you would change usage of `__LOG_LEVEL__` in the source helm chart (https://github.com/istio/istio/blob/master/manifests/charts/istio-cni/templates/configmap-cni.yaml) into something like `{{ "{{ .logLevel }}" }}`. 

Your point about end users possibly having their own config templates that would break with this change is a good one. However, a transition to beta is a good time to have breaking user-facing changes. I'm not sure if we would ever need the additional capabilities that a true template would provide, but it seems like it may be handy.

We'd need to add docs for these special variables, as well. 

@jwendell @rlenglet thoughts?

Reply:
I'd prefer to keep the upgrade of this config map simple. So keep the format identical for now.
This may be done in a separate PR. Let's try to maintain feature parity and compatibility for this PR.

Reply:
Just remember that 1.7 branch cut will happen next Tuesday. So, if this is going to change in a separate PR, it's better to be before that day.

I'm ok with leaving this as is, to keep compatibility and change that for 1.8. Ideally if we could have both working at the same time, that would be great, thus we could deprecate current one in 1.7 and remove it in 1.8. But I agree that time is short.